lib/derive/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+extern crate proc_macro;
+
+use proc_macro2::{Span, TokenStream};
+use proc_macro_error::{abort, abort_call_site, proc_macro_error, set_dummy};
+use quote::{format_ident, quote, quote_spanned};
+use syn::{punctuated::Punctuated, spanned::Spanned, token::Comma, *};
+
+mod parse;
+
+use crate::parse::{ExportAttr, WasmerAttr};
+
+#[proc_macro_derive(WasmerEnv, attributes(wasmer))]
+#[proc_macro_error]
+pub fn derive_wasmer_env(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input: DeriveInput = syn::parse(input).unwrap();
+    let gen = impl_wasmer_env(&input);
+    gen.into()
+}
+
+fn impl_wasmer_env_for_struct(
+    name: &Ident,
+    data: &DataStruct,
+    _attrs: &[Attribute],
+) -> TokenStream {
+    let (trait_methods, helper_methods) = derive_struct_fields(data);
+    quote! {
+        impl ::wasmer::WasmerEnv for #name {
```

Comment:
How do manage lifetime for `struct #name`? Like `struct<'a> S<'a> { … }`? It is OK to ignore this for the moment, but it can become really complex.

Reply:
We don't at all, I'm hoping to avoid that.

I posted on Slack about wanting to remove the lifetime from `NativeFunc` and friends because I don't think we're actually using it in practice and it is somewhat of a blocker to making this work

Reply:
I added lifetime support to the proc macro a bit ago, I think things should hopefully "just work"